internal/backend/grpc/dependency_graph.go

```diff
@@ -0,0 +1,243 @@
+	graph, err := s.runtime.GetDependencyGraph()
+	if err != nil {
+		return nil, err
+	}
+
+	filterControllers := xslices.ToSet(req.Controllers)
+
+	res := &resources.DependencyGraphResponse{}
+
+	nodesMap := map[string]string{}
+
+	genNodeID := func(label string) {
+		id := fmt.Sprintf("n_%s", label)
+
+		nodesMap[label] = id
+	}
+
+	addNode := func(edge controller.DependencyEdge, t resources.DependencyGraphResponse_Node_Type) error {
+		var label string
+
+		switch t {
+		case resources.DependencyGraphResponse_Node_RESOURCE:
+			label = edge.ResourceType
+		case resources.DependencyGraphResponse_Node_CONTROLLER:
+			label = edge.ControllerName
+		case resources.DependencyGraphResponse_Node_UNKNOWN:
+			return errors.New("unknown node type")
+		}
+
+		id, ok := nodesMap[label]
+
+		if !ok {
+			return nil
+		}
+
+		var (
+			labels []string
+			fields []string
+		)
+
+		if t == resources.DependencyGraphResponse_Node_RESOURCE {
+			ctx := actor.MarkContextAsInternalActor(ctx)
+
+			var resources []resource.Resource
+
+			if edge.ResourceID != "" {
+				res, err := s.state.Get(ctx, resource.NewMetadata(edge.ResourceNamespace, edge.ResourceType, edge.ResourceID, resource.VersionUndefined))
+				if err != nil && !state.IsNotFoundError(err) {
+					return err
+				}
+
+				if err == nil {
+					resources = append(resources, res)
+				}
+			} else {
+				response, err := s.state.List(ctx, resource.NewMetadata(edge.ResourceNamespace, edge.ResourceType, "", resource.VersionUndefined))
+				if err != nil {
+					return err
+				}
+
+				resources = response.Items
+			}
+
+			labelsMap := map[string]struct{}{}
+
+			for _, res := range resources {
+				for _, key := range res.Metadata().Labels().Keys() {
+					labelsMap[key] = struct{}{}
+				}
+			}
+
+			labels = slices.Sorted(maps.Keys(labelsMap))
+
+			res, err := protobuf.CreateResource(edge.ResourceType)
+			if err != nil {
+				return err
+			}
+
+			if reflect.ValueOf(res.Spec()).Kind() == reflect.Ptr {
+				err = json.Unmarshal([]byte("{}"), res.Spec())
+				if err != nil {
+					return err
+				}
+
+				type protobufWrapper interface {
+					GetValue() proto.Message
+				}
+
+				if message, ok := res.Spec().(protobufWrapper); ok {
+					v := reflect.TypeOf(message.GetValue()).Elem()
+
+					for i := range v.NumField() {
+						field := v.Field(i)
+
+						f := field.Tag.Get("json")
+						if f == "" {
+							continue
+						}
+
+						fields = append(fields, strings.Split(f, ",")[0])
+					}
+				}
+			}
+		}
+
+		res.Nodes = append(res.Nodes, &resources.DependencyGraphResponse_Node{
+			Id:     id,
+			Label:  label,
+			Type:   t,
+			Labels: labels,
+			Fields: fields,
+		})
+
+		return nil
+	}
+
+	addEdge := func(edge controller.DependencyEdge) error {
+		var (
+			source string
+			target string
+			style  string
+		)
+
+		switch edge.EdgeType {
+		case controller.EdgeOutputExclusive:
+			source = nodesMap[edge.ControllerName]
+			target = nodesMap[edge.ResourceType]
+			style = "bold"
```

Comment:
This is more of a nit, but should the backend tell how it should be rendered? Maybe instead we return the edge type as-is, and frontend can apply the style it wants. Backend would just return the data, no style.